servicetalk-encoding-api/src/main/java/io/servicetalk/encoding/api/ContentCodec.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright © 2020 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.encoding.api;
+
+import io.servicetalk.buffer.api.Buffer;
+import io.servicetalk.buffer.api.BufferAllocator;
+import io.servicetalk.concurrent.api.Publisher;
+
+/**
+ * API to support encode and decode of {@link Buffer}s.
+ * <p>
+ * <em>Implementations must provide thread safety semantics, since instances could be shared across threads.</em>
+ */
+public interface ContentCodec {
+
+    /**
+     * A unique textual representation for the coding.
+     *
+     * @return a unique textual representation for the coding.
+     */
+    CharSequence name();
+
+    /**
+     * Take a {@link Buffer} and encode its contents resulting in a {@link Buffer} with the encoded contents.
+     *
+     * @param src the {@link Buffer} to encode
+     * @param allocator the {@link BufferAllocator} to use for allocating auxiliary buffers or the returned buffer
+     * @return {@link Buffer} the result buffer with the content encoded
+     */
+    default Buffer encode(Buffer src, BufferAllocator allocator) {
+        return encode(src, src.readerIndex(), src.readableBytes(), allocator);
+    }
+
+    /**
+     * Take a {@link Buffer} and encode its contents resulting in a {@link Buffer} with the encoded contents.
+     *
+     * @param src the {@link Buffer} to encode
+     * @param offset the offset of the source to start reading from
+     * @param length the total length available for reading
+     * @param allocator the {@link BufferAllocator} to use for allocating auxiliary buffers or the returned buffer
+     * @return {@link Buffer} the result buffer with the content encoded
+     */
+    Buffer encode(Buffer src, int offset, int length, BufferAllocator allocator);
```

Comment:
Agreed, the offset part isn't needed, but the length it is in some cases to avoid creating view or slice.
Can we keep the changes as part of this PR https://github.com/apple/servicetalk/pull/1198/files#diff-0884917fb1a61a9cdef9580cd27730bd0d9863d6c93e5138b29866276576829bR55 ?